pkg/federate/kubefed/cluster.go

```diff
@@ -1,9 +1,314 @@
+	if !tokenFound || len(token) == 0 {
+		return nil, fmt.Errorf("the secret %#v is missing a non-empty value for %q", secret, corev1.ServiceAccountTokenKey)
+	}
+
+	kubeConfig, err := clientcmd.BuildConfigFromFlags(apiEndpoint, "")
+	if err != nil {
+		return nil, err
+	}
+
+	kubeConfig.CAData = kubeFedCluster.Spec.CABundle
+	kubeConfig.BearerToken = string(token)
+
+	return kubeConfig, nil
+}
+
+func (f *federator) initKubeFedClusterInformer(listerWatcher cache.ListerWatcher) {
+	// Providing 0 duration to an informer indicates that resync should be delayed as long as possible
+	resyncPeriod := 0 * time.Second
+	_, informer := cache.NewInformer(listerWatcher, &fedv1.KubeFedCluster{}, resyncPeriod, f)
+	f.kubeFedClusterInformer = informer
+}
+
+func newKubeFedClusterListerWatcher(kubeFedConfig *rest.Config) (cache.ListerWatcher, error) {
+	var obj runtime.Object = &fedv1.KubeFedCluster{}
+
+	gvk, err := apiutil.GVKForObject(obj, scheme.Scheme)
+	if err != nil {
+		return nil, fmt.Errorf("error finding GroupVersionKind for %T: %v", obj, err)
+	}
+
+	klog.V(3).Infof("GroupVersionKind for %T: %#v", obj, gvk)
+
+	mapper, err := apiutil.NewDiscoveryRESTMapper(kubeFedConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error creating RESTMapper from config: %v", err)
+	}
+
+	mapping, err := mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+	if err != nil {
+		return nil, fmt.Errorf("error creating RESTMapping for %#v: %v", gvk, err)
+	}
+
+	client, err := apiutil.RESTClientForGVK(gvk, kubeFedConfig, scheme.Codecs)
+	if err != nil {
+		return nil, fmt.Errorf("error creating RESTClient for %#v: %v", gvk, err)
+	}
+
+	return &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			result := &fedv1.KubeFedClusterList{}
+			err := client.Get().Namespace(kubeFedNamespace).Resource(mapping.Resource.Resource).
+				VersionedParams(&options, scheme.ParameterCodec).Do().Into(result)
+			return result, err
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			options.Watch = true
+			return client.Get().Namespace(kubeFedNamespace).Resource(mapping.Resource.Resource).
+				VersionedParams(&options, scheme.ParameterCodec).Watch()
+		},
+	}, nil
+}
+
+func (c *clusterWatcher) onAdd(clusterID string, kubeConfig *rest.Config) {
+	c.enqueueEvent(func() {
+		c.handler.OnAdd(clusterID, kubeConfig)
+	}, clusterID, "Add", defaultEnqueueTimerInterval)
+}
+
+func (c *clusterWatcher) onUpdate(clusterID string, kubeConfig *rest.Config) {
+	c.enqueueEvent(func() {
+		c.handler.OnUpdate(clusterID, kubeConfig)
+	}, clusterID, "Update", defaultEnqueueTimerInterval)
+}
+
+func (c *clusterWatcher) onRemove(clusterID string) {
+	c.enqueueEvent(func() {
+		c.handler.OnRemove(clusterID)
+	}, clusterID, "Remove", defaultEnqueueTimerInterval)
+}
+
+func (c *clusterWatcher) enqueueEvent(event func(), clusterID string, eventType string, enqueueTimerInterval time.Duration) {
```

Comment:
another lost comment after the push...

This enqueue implementation is not needed and I'm sure there is a structure that already does the same without us having to reimplement the wheel.

Reply:
We can switch the implementation later for something else which fits if we find it, like the work queue or something similar.